project_RL/sarsa/sarsa_lambda_agent.py

```diff
@@ -24,33 +24,80 @@ def init_q_value_table(self):
         """
-        self.q_value_table = np.zeros((self.state_size, self.action_size), np.int)
+        # self.q_value_table = np.zeros((self.state_size, self.action_size), np.int)
+        self.q_value_table = dd(lambda: dd(float))
 
     def init_eligibility_table(self):
         """Initialise eligibility trace table with zeros. Must be invoked before each episode.
         Its first dimension is the state size and the second dimension is the action size.
         """
-        self.eligibility_table = np.zeros((self.state_size, self.action_size))
+        # self.eligibility_table = np.zeros((self.state_size, self.action_size))
+        self.eligibility_table = dd(lambda: dd(float))
 
     def get_new_action_e_greedly(self, state):
         """With probability 1 - epsilon choose the greedy action.
         With probability epsilon choose random action.
         """
-        q_state = self.q_value_table[state["direction"]]
+        # q_state = self.q_value_table[state["direction"]]
+        q_state = self.q_value_table[tuple(state)]
         if random.random() < self.epsilon:
             return random.choice(range(self.action_size))
         else:
-            return np.argmax(q_state)
+            max_action = 0
+            max_val = -float('inf')
+            for action in q_state.keys():
+                if q_state[action] > max_val:
+                    max_val = q_state[action]
+                    max_action = action
+            return max_action
+
+    def get_new_action_greedly(self, state):
+        """With probability 1.0 choose the greedy action.
+        With probability 0.0 choose random action.
+        """
+        q_state = self.q_value_table[tuple(state)]
+        max_action = 0
+        max_val = -float('inf')
+        for action in q_state.keys():
+            if q_state[action] > max_val:
+                max_val = q_state[action]
+                max_action = action
+        return max_action
+
+    # def _get_goal_position(self, state):
+    #     grid = state['image']
+    #     rows = len(grid)
+    #     columns = len(grid[0])
+    #     for row in range(rows):
+    #         for column in range(columns):
+    #             if grid[row][column][0] == 8:
+    #                 return 10 * row + column
+    #     return 0
 
     def update(self, state, action, reward, new_state, new_action, done):
         """Updates the state action value for every pair state and action
         in proportion to TD-error and eligibility trace
 
         """
-        q_value_state_s = self.q_value_table[state["direction"]]
-        q_value_new_state = self.q_value_table[new_state["direction"]]
+        # import pdb; pdb.set_trace()
+        # q_value_state_s = self.q_value_table[tuple(state["direction"])]
+        # q_value_new_state = self.q_value_table[tuple(new_state["direction"])]
+        _state = tuple(state)
+        _new_state = tuple(new_state)
+        q_value_state_s = self.q_value_table[_state]
```

Comment:
I ran the code for many episodes. Looks like q_value_table length is always 1.
It seems the algorithm is learning the best action regardless of the state. That's why sometimes it is still random.

I debugged a little bit, the _state value is alway equal to this string '(\'image\', \'direction\', \'mission\')'
For me we can merge this code and I can create a new PR to fix it.